File: java/src/test/java/money_problem/domain/BankTest.java
package money_problem.domain;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static money_problem.domain.Currency.*;
import static money_problem.domain.DomainUtility.*;
import static org.assertj.vavr.api.VavrAssertions.assertThat;
import static org.junit.jupiter.params.provider.Arguments.of;

class BankTest {
    public static final Currency PIVOT_CURRENCY = EUR;
    private final Bank bank = Bank.withPivotCurrency(PIVOT_CURRENCY);

    private static Stream<Arguments> examplesOfDirectConversion() {
        return Stream.of(
                of(dollars(87), EUR, euros(72.5)),
                of(koreanWons(1_009_765), EUR, euros(751.313244047619)),
                of(euros(10), USD, dollars(12)),
                of(euros(543.98), USD, dollars(652.776))
        );
    }

    @ParameterizedTest
    @MethodSource("examplesOfDirectConversion")
    void convertDirectly(Money money, Currency to, Money expectedResult) {
        assertConversion(money, to, expectedResult);
    }

    private static Stream<Arguments> examplesOfConversionThroughPivotCurrency() {
        return Stream.of(
                of(dollars(10), KRW, koreanWons(11200)),
                of(dollars(-1), KRW, koreanWons(-1120)),
                of(koreanWons(39_345.50), USD, dollars(35.129910714285714)),
                of(koreanWons(1000), USD, dollars(0.8928571428571427))
        );
    }

    @ParameterizedTest
    @MethodSource("examplesOfConversionThroughPivotCurrency")
    void convertThroughPivotCurrency(Money money, Currency to, Money expectedResult) {
        assertConversion(money, to, expectedResult);
    }

    private void assertConversion(Money money, Currency to, Money expectedResult) {
        assertThat(bank.add(rateFor(1.2, USD))
                .flatMap(b -> b.add(rateFor(1344, KRW)))
                .flatMap(newBank -> newBank.convert(money, to)))
                .containsOnRight(expectedResult);
    }
}

File: c#/money-problem.Tests/BankTest.cs
using System.Collections.Generic;
using FluentAssertions.LanguageExt;
using money_problem.Domain;
using Xunit;

namespace money_problem.Tests;

public class BankTest
{
    private const Currency PivotCurrency = Currency.EUR;
    private readonly Bank bank;

    public BankTest() => this.bank = Bank.WithPivotCurrency(PivotCurrency);

    public static IEnumerable<object[]> ExamplesForConvertThroughPivotCurrency =>
        new List<object[]>
        {
            new object[] {10d.Dollars(), Currency.KRW, 11200d.KoreanWons()},
            new object[] {(-1d).Dollars(), Currency.KRW, (-1120d).KoreanWons()},
            new object[] {1000d.KoreanWons(), Currency.USD, 0.8928571428571427.Dollars()},
        };

    public static IEnumerable<object[]> ExamplesForConvertThroughExchangeRate =>
        new List<object[]>
        {
            new object[] {87d.Dollars(), Currency.EUR, 72.5d.Euros()},
            new object[] {1009765d.KoreanWons(), Currency.EUR, 751.313244047619d.Euros()},
            new object[] {10d.Euros(), Currency.USD, 12d.Dollars()},
            new object[] {10d.Euros(), Currency.KRW, 13440d.KoreanWons()},
        };

    [Theory]
    [MemberData(nameof(ExamplesForConvertThroughExchangeRate))]
    public void ConvertConvertThroughExchangeRate(Money money, Currency currency, Money expected) =>
        DomainUtility.WithExchangeRates(
                this.bank,
                DomainUtility.CreateExchangeRate(Currency.USD, 1.2),
                DomainUtility.CreateExchangeRate(Currency.KRW, 1344))
            .Convert(money, currency)
            .Should()
            .Be(expected);

    [Fact]
    public void ConvertInDollarsFromEurosWithUpdatedRate() =>
        DomainUtility.WithExchangeRates(
                this.bank,
                DomainUtility.CreateExchangeRate(Currency.USD, 1.1),
                DomainUtility.CreateExchangeRate(Currency.USD, 1.2))
            .Convert(10d.Euros(), Currency.USD)
            .Should()
            .Be(12d.Dollars());

    [Theory]
    [MemberData(nameof(ExamplesForConvertThroughPivotCurrency))]
    public void ConvertThroughPivotCurrency(Money money, Currency currency, Money expected) =>
        DomainUtility.WithExchangeRates(
                this.bank,
                DomainUtility.CreateExchangeRate(Currency.USD, 1.2),
                DomainUtility.CreateExchangeRate(Currency.KRW, 1344))
            .Convert(money, currency)
            .Should()
            .Be(expected);
}